include/crocoddyl/multibody/actions/contact-fwddyn.hxx

```diff
@@ -103,6 +103,28 @@ void DifferentialActionModelContactFwdDynamicsTpl<Scalar>::calc(
   d->cost = d->costs->cost;
 }
 
+template <typename Scalar>
+void DifferentialActionModelContactFwdDynamicsTpl<Scalar>::calc(
+    const boost::shared_ptr<DifferentialActionDataAbstract>& data, const Eigen::Ref<const VectorXs>& x) {
+  if (static_cast<std::size_t>(x.size()) != state_->get_nx()) {
+    throw_pretty("Invalid argument: "
+                 << "x has wrong dimension (it should be " + std::to_string(state_->get_nx()) + ")");
+  }
+
+  Data* d = static_cast<Data*>(data.get());
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> q = x.head(state_->get_nq());
+  const Eigen::VectorBlock<const Eigen::Ref<const VectorXs>, Eigen::Dynamic> v = x.tail(state_->get_nv());
+
+  pinocchio::computeAllTerms(pinocchio_, d->pinocchio, q, v);
```

Comment:
This probably computes more than needed because it also computes some dynamics quantities such as M and h.

Reply:
Hey @andreadelprete 

Your observation is right, but it is still not convenient for predictive control application. Let me explain this claim briefly.

In MPC applications, you might receded the running model. So, the terminal node will become the last running one and so on. Therefore, we will need to compute all the terms as we reuse the `calc` results (performed in the forward pass) in the next solver iteration: https://github.com/loco-3d/crocoddyl/blob/devel/src/core/solvers/ddp.cpp#L127

(Note that we don't run `calc` before `calcDiff`)

Due to this justification, I won't follow this suggestion. We could open a discussion on this topic, if you think we need to dig more on it.

Reply:
It's fine by me. In any case it's a very small computation time.